codegenerator/cli/templates/static/codegen/src/eventFetching/FetchState.res

```diff
@@ -319,60 +330,183 @@ let rec updateInternal = (
-  | (DynamicContractRegister(dynamicContractId, nextRegistered), id) =>
-    nextRegistered->updateInternal(
+  | (DynamicContractRegister({id: dynamicContractId, nextRegister}), id) =>
+    nextRegister->updateInternal(
       ~id,
       ~latestFetchedBlock,
       ~reversedNewItems,
-      ~isFetchingAtHead,
       ~parent=register->Parent.make(~dynamicContractId, ~parent),
     )
   | (RootRegister(_), DynamicContract(_)) => Error(UnexpectedRegisterDoesNotExist(id))
   }
 }
 
+/**
+Inserts a dynamic contract register to the head of a given
+register. It will then precede the given register in the chain
+*/
+let addNewRegisterToHead = (
+  self,
+  ~registeringEventBlockNumber,
+  ~registeringEventLogIndex,
+  ~contractAddressMapping,
+) => {
+  let id: dynamicContractId = {
+    blockNumber: registeringEventBlockNumber,
+    logIndex: registeringEventLogIndex,
+  }
+  let registerType = DynamicContractRegister({id, nextRegister: self})
+
+  let dynamicContracts =
+    DynamicContractsMap.empty->DynamicContractsMap.add(
+      id,
+      contractAddressMapping->ContractAddressingMap.getAllAddresses,
+    )
+
+  {
+    registerType,
+    latestFetchedBlock: {
+      blockNumber: registeringEventBlockNumber - 1,
+      blockTimestamp: 0,
+    },
+    contractAddressMapping,
+    dynamicContracts,
+    fetchedEventQueue: [],
+    firstEventBlockNumber: None,
+  }
+}
+
+/**
+Adds a new dynamic contract registration. It inserts the registration ordered in the
+chain from earliest registered contract to latest. So if this is being called on a batch
+of registrations its best to do this in order of latest to earliest to reduce recursions
+of this function.
+*/
+let rec addDynamicContractRegister = (
+  self: register,
+  ~registeringEventBlockNumber,
+  ~registeringEventLogIndex,
+  ~dynamicContractRegistrations: array<TablesStatic.DynamicContractRegistry.t>,
+  ~parent: option<Parent.t>=?,
+) => {
+  let handleParent = updated =>
+    switch parent {
+    | Some(parent) => parent->Parent.joinChild(updated)
+    | None => updated
+    }
+
+  let addToHead = updated =>
+    updated
+    ->addNewRegisterToHead(
+      ~contractAddressMapping=dynamicContractRegistrations
+      ->Array.map(d => (d.contractAddress, (d.contractType :> string)))
+      ->ContractAddressingMap.fromArray,
+      ~registeringEventLogIndex,
+      ~registeringEventBlockNumber,
+    )
+    ->handleParent
+
+  let latestFetchedBlockNumber = registeringEventBlockNumber - 1
+
+  switch self.registerType {
+  | RootRegister(_) => self->addToHead
+  | DynamicContractRegister(_) if latestFetchedBlockNumber <= self.latestFetchedBlock.blockNumber =>
+    self->addToHead
+  | DynamicContractRegister({id: dynamicContractId, nextRegister}) =>
+    nextRegister->addDynamicContractRegister(
+      ~registeringEventBlockNumber,
+      ~registeringEventLogIndex,
+      ~dynamicContractRegistrations,
+      ~parent=self->Parent.make(~dynamicContractId, ~parent),
+    )
+  }
+}
```

Comment:
These two functions were simply moved to be accessible to the functions below. No logic changes.